### ISSUE TITLE
API mocking with MSW fails after upgrade to miniflare 2.2.0

### ISSUE THREAD

**Issue by fbarthez** [cloudflare/miniflare#162]
Hey @mrbbot, thanks for all the work on miniflare, it's been a game-changer!
With miniflare 1, I've been using [mock service worker](https://mswjs.io/) to mock API responses. When attempting to upgrade to miniflare 2.2.0, my tests failed. I've no idea where to start looking for a solution, so I've put up [a fairly minimal demo](https://github.com/fbarthez/miniflare-msw/), hoping that'll make it easier to pinpoint the cause. HEAD is using miniflare 2.2.0, while the [miniflare-1.4.1 branch](https://github.com/fbarthez/miniflare-msw/tree/miniflare-1.4.1) shows what it looks like when it's working. Running `npm test` should show failing tests for HEAD and passing for the latter.

**Comment by mrbbot** [comment]
Hey! 👋 I think this is because instead of `node-fetch`, Miniflare 2 uses [`undici`](https://github.com/nodejs/undici) for its `fetch` implementation, which doesn't use the built-in `http` module, so bypasses [Mock Service Worker's monkeypatching](https://mswjs.io/docs/comparison#nock).

I'm not sure if this is you, but a similar question was recently asked on the Remix Discord server. The response there might be useful at least until Mock Service Worker is updated to work with `undici`: https://discord.com/channels/770287896669978684/777863813981536280/935974246339985438

**Comment by fbarthez** [comment]
Thanks! That'll get me back on track, probably using the [built-in mocking feature of undici](https://nodejs.medium.com/introducing-undici-4-1e321243e007#b30b) for now.

**Comment by SupremeTechnopriest** [comment]
Hey @fbarthez, 

Were you ever able to get this to work?  Im testing out the wrangler beta and Im not able to get the undici mocking to intercept fetch requests.

**Comment by fbarthez** [comment]
Hey @SupremeTechnopriest, no it's still on my todo list and will probably stay there for at least another week. ;/

**Comment by SupremeTechnopriest** [comment]
I kind of need to be able to mock the endpoints.  I also really like the MSW API.  I might bite the bullet and write the interceptor, but it might also be nice for `jest-environment-miniflare` to expose its own mocking framework.  Maybe a wrapper around the undici ones.  @mrbbot what do you think about starting a discussion around that?

**Comment by SupremeTechnopriest** [comment]
Seems like there are multiple modules that have undici as a dependency... `miniflare`, `@minflare/core`, `@miniflare/http-server` etc... Not sure which one is the one actually performing the request.  I'm pretty sure you need to use the exported mocking functions from the instance of undici doing the request.  The below test fails no matter where I import undici from. 

```ts
import { MockAgent, setGlobalDispatcher } from '../../node_modules/@miniflare/core/node_modules/undici'

const agent = new MockAgent()
agent.disableNetConnect()

agent.get('http://foo.com')
  .intercept({
    path: '/',
    method: 'GET'
  })
  .reply(200, 'bar')

setGlobalDispatcher(agent)

describe('undiciMock', () => {
  test('Mocks response', async () => {
    const request = new Request('http://foo.com/')
    const res = await fetch(request)
    const text = await res?.text()
    expect(text).toBe('bar')
  })
})
```

If you have any insight as to what instance of undici we should be mocking that would be very helpful @mrbbot.  Still think we should expose the mocking features of undici more cleanly. 

Using `jest-environment-miniflare` with the default config.

**Comment by mrbbot** [comment]
Hey @SupremeTechnopriest! 👋 Unfortunately, you won't be able to use `setGlobalDispatcher` inside `jest-environment-miniflare`. Miniflare's `undici` version is imported *outside* the Jest environment in Node. Jest and Node don't share a module cache, so trying to import `undici` inside Jest will always give you a different version.

Currently, I think the best option for `jest-environment-miniflare` is to use Jest's function mocks to mock the global `fetch` function. Something like... *(untested 😅)*

```js
const fetchSpy = jest.spyOn(globalThis, "fetch");
fetchSpy.mockImplementation((input, init) => {
  const req = new Request(input, init);
  const url = new URL(req.url);
  if (req.method === "GET" && url.pathname === "/") {
    return new Response("bar");
  }
  return new Response(null, { status: 404 });
});

...

jest.restoreAllMocks();
```

I'd definitiely like to start a discussion on this though. Maybe a global API like `getMiniflareMockAgent()` that returned a correctly-setup `MockAgent`?

**Comment by SupremeTechnopriest** [comment]
@mrbbot,

That's exactly what I was thinking! Miniflare has made my life soooooooo much better and this would be the icing on the cake.  Let me know how I can help!

**Comment by SupremeTechnopriest** [comment]
Your fetch spy works by the way @mrbbot.

**Comment by SupremeTechnopriest** [comment]
Wrote a little warpper while we sort out the mock global.

```ts
export function MockFetch (base: string) {
  const mocks: IFetchMock[] = []
  const fetchSpy = jest.spyOn(globalThis, 'fetch')
  fetchSpy.mockImplementation(async (input: string | Request, init?: Request | RequestInit | undefined): Promise<Response> => {
    const request = new Request(input, init)

    for (const mock of mocks) {
      if (`${base}${mock.path}` === request.url && request.method === mock.method) {
        const resInit = {
          status: mock.status,
          statusText: mock.statusText,
          headers: {}
        }
        if (mock.headers) resInit.headers = mock.headers
        return new Response(mock.body, resInit)
      }
    }
    return new Response(null, { status: 404, statusText: 'Not Found' })
  })

  const intercept = (mock: IFetchMock) => {
    mocks.push(mock)
  }

  const restore = () => {
    mocks.length = 0
    jest.restoreAllMocks()
  }

  return { intercept, restore }
}
```

```ts
const { intercept, restore } = MockFetch('http://foo.com')

intercept({
  path: '/',
  method: 'GET',
  status: 200,
  statusText: 'OK',
  body: 'bar',
  headers: {
    'X-Foo': 'bar'
  }
})

afterAll(restore)

test('Mock', async () => {
  const request = new Request('http://foo.com/')
  const res = await fetch(request)
  const text = await res?.text()
  expect(text).toBe('bar')
  expect(res.headers.get('x-foo')).toBe('bar')
})
```

**Comment by ptim** [comment]
This was super helpful to me, as [jest-fetch-mock](https://github.com/jefflau/jest-fetch-mock/issues/218#issuecomment-1078606077) doesn't return actual responses!

Here's the wrapper I'm using now:

```ts
// @test/helpers/MockAgent.ts

type Route = {
  matcher: (input: Request | string, init?: RequestInit | Request) => boolean
  response: Response
}

/**
 * Returns a handler function composed of passed Handlers,
 * falling back to a 404 Response on no match.
 * 
 * @see https://github.com/cloudflare/miniflare/issues/162#issuecomment-1059549646
 */
export function createRequestHandlers(routes: Route[] = []) {
  return async function handler(
    input: Request | string,
    init?: RequestInit | Request,
  ) {
    for (const { matcher, response } of routes) {
      if (matcher(input, init)) return response
    }

    return new Response('Not Found', { status: 404 })
  }
}

/**
 * Convenience function returns true if the GET request matches passed pathname
 */
export function createGetMatcher(pathname: string) {
  return function (input: Request | string, init?: RequestInit | Request) {
    const req = new Request(input, init)
    const url = new URL(req.url)
    return req.method === 'GET' && pathname === url.pathname
  }
}
```

```ts
import { jest } from '@jest/globals'
import {
  basicRequestHandler,
  createRequestHandlers,
  createGetMatcher as get,
} from '@test/helpers/MockAgent'

afterEach(() => {
  jest.restoreAllMocks()
})

test('should mock different routes', async () => {
  let mockBarOnce = true

  const requestHandlers = createRequestHandlers([
    {
      matcher: get('/foo'),
      response: new Response('foo'),
    },
    {
      matcher: (input, init) => {
        const req = new Request(input, init)
        const url = new URL(req.url)
        const matches = req.method === 'GET' && url.pathname === '/bar'
        if (matches && mockBarOnce) {
          mockBarOnce = false
          return true
        }
        return false
      },
      response: new Response('bar'),
    },
  ])

  const fetchSpy = jest.spyOn(globalThis, 'fetch')
  fetchSpy.mockImplementation(requestHandlers)

  let req = new Request('http://localhost/')
  let res = await fetch(req)
  expect(res.status).toBe(404)

  req = new Request('http://localhost/foo')
  res = await fetch(req)
  expect(res.status).toBe(200)
  expect(await res.text()).toBe('foo')

  // mock GET /bar only once
  req = new Request('http://localhost/bar')
  res = await fetch(req)
  expect(res.status).toBe(200)
  expect(await res.text()).toBe('bar')

  // second call to GET /bar passes through
  req = new Request('http://localhost/bar')
  res = await fetch(req)
  expect(res.status).toBe(404)

  expect(fetchSpy.mock.calls).toHaveLength(4)
})
```

**Comment by SupremeTechnopriest** [comment]
This looks nice! Good stop gap until the true undici mocks are exposed.

**Comment by aiji42** [comment]
I was hoping that perhaps this issue had been resolved with the release of 2.4.0 yesterday, but it has been postponed to 2.5.0.
I am very much looking forward to the next release 👀

**Comment by LukePammant** [comment]
@ptim and @SupremeTechnopriest if I'm not mistaken your mock `fetch` is only being used in the `jest` test code itself because you're calling fetch directly in the test. So any invokation to `miniflare` via something like `miniflare.dispatchFetch` will not use your mock. Is that right or am I understanding it correctly? 

I'd love to do a simple `spyOn` too but `miniflare` still uses the `undici` fetch since `jest` and `miniflare` don't share a module cache as @mrbbot pointed out.

**Comment by SupremeTechnopriest** [comment]
It should mock the fetch call for your module code as well.  Are you seeing something different?

**Comment by LukePammant** [comment]
@SupremeTechnopriest yeah, if I add a mock implementation that simply throws an exception and then call my worker via `miniflare.dispatchFetch` the request goes through and calls the endpoint. Example that should just fail but works fine:

``` ts
const fetchSpy = jest.spyOn(globalThis, 'fetch')
fetchSpy.mockImplementation(async (input: string | Request, init?: Request | RequestInit | undefined): Promise<Response> => {
    throw "No fetch for you"
})

it('shouldn throw', async () => {
     const miniflare = new Miniflare({
       // ... setup miniflare vars
     });
    miniflare.dispatchFetch("https://google.com");
});    
```

Maybe it's because I'm not using the new Modules syntax for CF Workers?

**Comment by msaiducar** [comment]
@SupremeTechnopriest and @LukePammant, I can confirm that mocking the `globalThis.fetch` doesn't work with `miniflare.dispatchFetch`. We have some `fetch()` calls to some external APIs inside our worker, we wanted to test those calls, but spying fetch was not possible. We used [this library](https://openbase.com/js/mock-http-server) to workaround. We start a local HTTP server and use this server instead of external APIs.

**Comment by ptim** [comment]
Works for me if I call the handler directly, but not with `mf.dispatchFetch`:

Works: 

```ts
// src/index.ts
export async function handleRequest(request: Request, env: Bindings) {
  const res = await fetch("http://not-a-real-domain.foo");
  return res;
}
```

```ts
import { handleRequest } from "@/index";
// ...

test("should provide mockFetch", async () => {
  const _fetchSpy = mockFetch(
    "http://not-a-real-domain.foo",
    new Response("mocked")
  );

  const env = getMiniflareBindings();
  const req = new Request("http://localhost/");
  const res = await handleRequest(req, env);

  expect(_fetchSpy).toHaveBeenCalledTimes(1);
  expect(res.status).toBe(200);
  expect(await res.text()).toBe("mocked");
});
```
